src/validator/is-ipv4-addr.ts

```diff
@@ -0,0 +1,67 @@
+import {STRule} from '../rule/rule';
+import {STRuleFn} from '../rule/fn';
+import {STRuleModifiers} from '../rule/modifiers';
+import {STRuleNode} from '../rule/node';
+import {STRuleNodeType} from '../rule/node-type';
+
+export type STOpIsIpv4Addr<CallerType> = () => CallerType;
+
+export const isIpv4Addr = (curr: string): boolean => {
+	//Always a string.
+	if (typeof curr !== 'string') {
+		return false;
+	}
+	//Is valid if it has whitespace which can be trimmed with .trim()
+	//Invalid if it contains any other whitespace which cannot be trimmed with .trim()
+	if (!curr.trim()) {
```

Comment:
Using trim is good, however, there is both a performance cost to using it (tiny) and a consistency issue (more important). Anytime we're using string operations 2 or more times, it's a good idea to save the result to avoid performing the check over again. 

Imagine trim was instead a very expensive string search `O(n^3)` for some reason.
This might be something like:
```typescript
if (!curr.doExpensiveSearch('searchTerm')) {
   return false;
}

const myItem = curr.doExpensiveSearch('searchTerm');
```

We eat the expensive search once to see if what we're looking for exists. If not, we bail out it was performed only once. If it's found we then perform it again and eat the cost twice. Bringing it back to this function, we save the trimmed result and then use it in the second part.

```typescript
const trimmed = curr.trim();
if (!trimmed) {
   return false;
}

const pieces = trimmed.split('.');
```

The second issue was we validate the trimmed string initially, but then do not split the trimmed result, we split the original result.